### PR TITLE
Prevent concurent state emit in iterable player when setting playback

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
When switching layouts using any of the experimental players the app no longer crashes with an error about emitting playerState before last playerState was rendered.

**Description**
When setting the playback speed for the player the player needs to emit state so downstream components can update with the new playback speed. The iterable player emits the state only when idle, however even when idle the player may be loading blocks and emitting state which would cause concurrent state emits which is invalid in our pipeline.

This change fixes this behavior by tracking whether a state emit is in-flight within the emitState method and waiting for any in-flight state to finish before emitting again.

Fixes: #3375

### CODE REVIEW

**Comment by defunctzombie** [comment]
@jtbandes rather than using debouncePromise (which I didn't like the behavior of since it doesn't allow me to await on the call that debouncePromise returns), I instead opted to make setPlaybackSpeed invoke the idle state to trigger a state update. I find this simpler to reason about given the state machine nature of this player.

**Review by jtbandes — APPROVED** [review]

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts [hunk 1]
```diff
@@ -222,10 +222,10 @@ export class IterablePlayer implements Player {
     this._speed = speed;
     this._metricsCollector.setSpeed(speed);
 
-    // If we are idling then we emit state to reflect the new speed in the player state.
-    // For other states, we let them emit their own state update
+    // If we are idling then we might not emit any new state so we use a state change to idle state
+    // to trigger an emit so listeners get updated with the new speed setting.
     if (this._state === "idle") {
-      void this._emitState();
+      this._setState("idle");
```
[the remark this answers is not in the document]
Comment (defunctzombie) [reply packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts:228]: Yea - thats why I added a comment. I'm not happy with it either but I also wasn't happy with debouncePromise somehow. I might come up with a new idea after thinking on this code more but want to land this fix until those ideas come to me.

**Comment by defunctzombie** [comment]
> It does, you can access the current promise via:

That's the current one but not the one that might run as a result of me calling the debounced function.

**Comment by jtbandes** [comment]
> That's the current one but not the one that might run as a result of me calling the debounced function.

What if there never is a run that directly results from your call? If there are multiple calls while one is pending only the latest one will be queued.

**Comment by defunctzombie** [comment]
> What if there never is a run that directly results from your call? If there are multiple calls while one is pending only the latest one will be queued.

I don't follow this specifically. All of this is kind of why I find the state transitioning approach to be easier to reason about.

**Comment by defunctzombie** [comment]
@esthersweon no need to add to changelog

**Pull request merged**